File: s3JavaSDKExample/sample/JAVA_SDK_Sample/src/main/java/hicloud/s3/sample/ACLSerialTesting.java
package hicloud.s3.sample;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.Writer;
import java.util.Properties;

import com.amazonaws.AmazonClientException;
import com.amazonaws.AmazonServiceException;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.model.AccessControlList;
import com.amazonaws.services.s3.model.BucketVersioningConfiguration;
import com.amazonaws.services.s3.model.CannedAccessControlList;
import com.amazonaws.services.s3.model.CanonicalGrantee;
import com.amazonaws.services.s3.model.EmailAddressGrantee;
import com.amazonaws.services.s3.model.GroupGrantee;
import com.amazonaws.services.s3.model.ListVersionsRequest;
import com.amazonaws.services.s3.model.Owner;
import com.amazonaws.services.s3.model.Permission;
import com.amazonaws.services.s3.model.PutObjectRequest;
import com.amazonaws.services.s3.model.PutObjectResult;
import com.amazonaws.services.s3.model.S3VersionSummary;
import com.amazonaws.services.s3.model.SetBucketVersioningConfigurationRequest;
import com.amazonaws.services.s3.model.VersionListing;

public class ACLSerialTesting {

  private static File createSampleFile() throws IOException {
    File file = File.createTempFile("aws-java-sdk-", ".txt");
    file.deleteOnExit();

    Writer writer = new OutputStreamWriter(new FileOutputStream(file));
    writer.write("abcdefghijklmnopqrstuvwxyz\n");
    writer.write("01234567890112345678901234\n");
    writer.write("!@#$%^&*()-=[]{};':',.<>/?\n");
    writer.write("01234567890112345678901234\n");
    writer.write("abcdefghijklmnopqrstuvwxyz\n");
    writer.close();

    return file;
  }

  private static void basicPutBucketACL(String bucketName) throws IOException {
    AmazonS3 s3 = S3Client.getClient();
    Properties prop = new Properties();
    prop.load(ACLSerialTesting.class.getClassLoader().getResourceAsStream("config.properties"));

    AccessControlList acl = new AccessControlList();
    Owner owner = new Owner();
    owner.setDisplayName(prop.getProperty("ownerDisplayName"));
    owner.setId(prop.getProperty("ownerUserID")); // User ID

    try {
      System.out.println("Creating bucket " + bucketName + "\n");
      s3.createBucket(bucketName);

      acl.setOwner(owner);
      acl.grantPermission(new CanonicalGrantee(prop.getProperty("ownerCanonicalID")),
          Permission.FullControl); // Canonical ID
      acl.grantPermission(GroupGrantee.AllUsers, Permission.Read);
      acl.grantPermission(new EmailAddressGrantee(prop.getProperty("userAEmail")),
          Permission.WriteAcp); // e-mail

      System.out.println("put bucket ACL");
      s3.setBucketAcl(bucketName, acl);

      System.out.println("put bucket ACL with Canned ACL");
      s3.setBucketAcl(bucketName, CannedAccessControlList.PublicRead);

      System.out.println("Tear down..");
      s3.deleteBucket(bucketName);

    } catch (AmazonServiceException ase) {
      System.out.println("Caught an AmazonServiceException, which means your request made it "
          + "to Amazon S3, but was rejected with an error response for some reason.");
      System.out.println("Error Message:    " + ase.getMessage());
      System.out.println("HTTP Status Code: " + ase.getStatusCode());
      System.out.println("AWS Error Code:   " + ase.getErrorCode());
      System.out.println("Error Type:       " + ase.getErrorType());
      System.out.println("Request ID:       " + ase.getRequestId());
    } catch (AmazonClientException ace) {
      System.out.println("Caught an AmazonClientException, which means the client encountered "
          + "a serious internal problem while trying to communicate with S3, "
          + "such as not being able to access the network.");
      System.out.println("Error Message: " + ace.getMessage());
    }
  }

  private static void vPutObjectACL(String bucketName, String fileName) throws IOException {
    Properties prop = new Properties();
    prop.load(ACLSerialTesting.class.getClassLoader().getResourceAsStream("config.properties"));

    AccessControlList acl = new AccessControlList();

    Owner owner = new Owner();
    owner.setDisplayName(prop.getProperty("ownerDisplayName"));
    owner.setId(prop.getProperty("ownerCanonicalID")); // Canonical ID

    AmazonS3 s3 = S3Client.getClient();

    try {
      System.out.println("Creating bucket " + bucketName + "\n");
      s3.createBucket(bucketName);

      // Enable versioning
      System.out.println("basic put bucket versioning");
      BucketVersioningConfiguration config = new BucketVersioningConfiguration();
      config.setStatus("Enabled");
      SetBucketVersioningConfigurationRequest version =
          new SetBucketVersioningConfigurationRequest(bucketName, config);
      s3.setBucketVersioningConfiguration(version);

      System.out.println("Uploading a new object to S3 from a file\n");
      PutObjectResult object =
          s3.putObject(new PutObjectRequest(bucketName, fileName, createSampleFile()));
      String vid = object.getVersionId();

      acl.setOwner(owner);
      acl.grantPermission(new CanonicalGrantee(prop.getProperty("userACanonicalID")),
          Permission.ReadAcp); // Canonical ID
      acl.grantPermission(GroupGrantee.AllUsers, Permission.Read);
      acl.grantPermission(new EmailAddressGrantee(prop.getProperty("userBEmail")),
          Permission.WriteAcp);
      acl.grantPermission(new CanonicalGrantee(prop.getProperty("ownerCanonicalID")),
          Permission.FullControl); // Canonical ID
      // set Object ACL
      System.out.println("Put object ACL...\n");
      s3.setObjectAcl(bucketName, fileName, vid, acl);

      // set Canned ACL for the object with specific version
      System.out.println("Put object ACL with canned acl + version id \n");
      s3.setObjectAcl(bucketName, fileName, vid, CannedAccessControlList.PublicReadWrite);


      System.out.println("Tear down..");
      // list all versions of the objects in the bucket
      VersionListing clear = s3.listVersions(new ListVersionsRequest().withBucketName(bucketName));
      for (S3VersionSummary s : clear.getVersionSummaries()) {
        System.out.println(s.getBucketName());
        System.out.println(s.getKey());
        System.out.println(s.getVersionId());
        // delete the specific version of a object
        s3.deleteVersion(s.getBucketName(), s.getKey(), s.getVersionId());
      }
      s3.deleteBucket(bucketName);
      System.out.println("DONE");

    } catch (AmazonServiceException ase) {
      System.out.println("Caught an AmazonServiceException, which means your request made it "
          + "to Amazon S3, but was rejected with an error response for some reason.");
      System.out.println("Error Message:    " + ase.getMessage());
      System.out.println("HTTP Status Code: " + ase.getStatusCode());
      System.out.println("AWS Error Code:   " + ase.getErrorCode());
      System.out.println("Error Type:       " + ase.getErrorType());
      System.out.println("Request ID:       " + ase.getRequestId());
    } catch (AmazonClientException ace) {
      System.out.println("Caught an AmazonClientException, which means the client encountered "
          + "a serious internal problem while trying to communicate with S3, "
          + "such as not being able to access the network.");
      System.out.println("Error Message: " + ace.getMessage());
    }
  }


  private static void basicPutObjectACL(String bucketName, String fileName) throws IOException {
    Properties prop = new Properties();
    prop.load(ACLSerialTesting.class.getClassLoader().getResourceAsStream("config.properties"));

    AccessControlList acl = new AccessControlList();

    Owner owner = new Owner();
    owner.setDisplayName(prop.getProperty("ownerDisplayName"));
    owner.setId(prop.getProperty("ownerCanonicalID")); // Canonical ID


    AmazonS3 s3 = S3Client.getClient();

    try {
      System.out.println("Creating bucket " + bucketName + "\n");
      s3.createBucket(bucketName);

      System.out.println("Uploading a new object to S3 from a file\n");
      s3.putObject(new PutObjectRequest(bucketName, fileName, createSampleFile()));

      acl.setOwner(owner);
      acl.grantPermission(new CanonicalGrantee(prop.getProperty("userACanonicalID")),
          Permission.ReadAcp); // Canonical ID
      acl.grantPermission(GroupGrantee.AllUsers, Permission.Read);
      acl.grantPermission(new EmailAddressGrantee(prop.getProperty("userBEmail")),
          Permission.WriteAcp);
      acl.grantPermission(new CanonicalGrantee(prop.getProperty("ownerCanonicalID")),
          Permission.FullControl); // Canonical ID

      System.out.println("Put object ACL...\n");
      s3.setObjectAcl(bucketName, fileName, acl);


      System.out.println("Put object ACL with canned acl...\n");
      s3.setObjectAcl(bucketName, fileName, CannedAccessControlList.PublicReadWrite);

      System.out.println("Tear down..");

      s3.deleteObject(bucketName, fileName);

      s3.deleteBucket(bucketName);

    } catch (AmazonServiceException ase) {
      System.out.println("Caught an AmazonServiceException, which means your request made it "
          + "to Amazon S3, but was rejected with an error response for some reason.");
      System.out.println("Error Message:    " + ase.getMessage());
      System.out.println("HTTP Status Code: " + ase.getStatusCode());
      System.out.println("AWS Error Code:   " + ase.getErrorCode());
      System.out.println("Error Type:       " + ase.getErrorType());
      System.out.println("Request ID:       " + ase.getRequestId());
    } catch (AmazonClientException ace) {
      System.out.println("Caught an AmazonClientException, which means the client encountered "
          + "a serious internal problem while trying to communicate with S3, "
          + "such as not being able to access the network.");
      System.out.println("Error Message: " + ace.getMessage());
    }
  }

  public static void main(String args[]) throws IOException {
    System.out.println("hello world");

    /*
     * test 1. PutBucketACL with user defined ACL 2. PutBucketACL with canned ACL 3. GetBucketACL
     */
    basicPutBucketACL(args[0]);

    /*
     * test 1. PutObjectACL with user defined ACL 2. PutObjectACL with canned ACL 3. GetObjectACL
     */
    basicPutObjectACL(args[0], args[1]);

    /*
     * test 1. PutObjectACL with user defined ACL with vid 2. PutObjectACL with canned ACL with vid
     * 3. GetObjectACL with vid
     */
    vPutObjectACL(args[0], args[1]);

  }

}

File: s3PhpSDKExample/sample/ACLSerialTesting.php
<?php

use Aws\S3\Enum\CannedAcl;
use Aws\S3\Exception\S3Exception;

require 'client.php';

function createSampleFile()
{
    $temp = tmpfile();
    $content = "abcdefghijklmnopqrstuvwxyz\n01234567890112345678901234\n!@#$%^&*()-=[]{};':',.<>/?<br>01234567890112345678901234\nabcdefghijklmnopqrstuvwxyz\n";
    fwrite($temp, $content);
    fseek($temp, 0);
    return $temp;
    fclose($temp); // this removes the file
}

function createBucket($bucketname, $loc=null)
{
    global $client;
    try {
        $result = $client->createBucket(array(
                'Bucket' => $bucketname ,
                'LocationConstraint' => $loc
        ));
    } catch (BucketAlreadyExistsException $e) {
        echo 'Bucket "'. $bucketname .'" already exists! <br>' .$e->getMessage()."<br><br>" ;
    } catch (S3Exception $e) {
        echo "Caught an AmazonServiceException, which means your request made it to Amazon S3, but was rejected with an error response for some reason.";
        echo "Error Message:    " . $e->getMessage()."<br>";
        echo "HTTP Status Code: " . $e->getStatusCode()."<br>";
        echo "AWS Error Code:   " . $e->getExceptionCode()."<br>";
        echo "Error Type:       " . $e->getExceptionType()."<br>";
        echo "Request ID:       " . $e->getRequestId()."<br>";
    }
}

function putObject($bucketname, $objName, $tmpfile, $ACL = CannedAcl::PRIVATE_ACCESS)
{
    global $client;
    $result = $client->putObject(array(
            'Bucket' => $bucketname,
            'Key'    => $objName,
            'Body'   => $tmpfile,
            'ACL'    => $ACL
    ));
}

function showObjectACL($bucketname, $file)
{
    global $client;
    $result=$client->getObjectAcl(array(
            'Bucket' => $bucketname,
            'Key' => $file
    ));
    echo "\n";
    echo "<br><br>Owner DisplayName: " . $result['Owner']['DisplayName'] . "<br>";
    echo "Owner ID: " . $result['Owner']['ID'] . "<br>";
    foreach ($result['Grants'] as $Grantee) {
        if (! empty($Grantee['Grantee']['Type'])) {
            echo  "Grantee type: " . $Grantee['Grantee']['Type'] . "<br>";
        }
        if (! empty($Grantee['Grantee']['ID'])) {
            echo  "Grantee ID: " . $Grantee['Grantee']['ID'] ."<br>";
        }
        if (! empty($Grantee['Grantee']['URI'])) {
            echo  "Grantee URI: " . $Grantee['Grantee']['URI'] ."<br>";
        }
        if (! empty($Grantee['Grantee']['DisplayName'])) {
            echo  "Grantee DisplayName: " . $Grantee['Grantee']['DisplayName'] . "<br>";
        }
        if (! empty($Grantee['Permission'])) {
            echo  "Grantee Permission: " . $Grantee['Permission']."<br><br>";
        }
    }
    echo "\n";
}

function showBucketACL($bucketname)
{
    global $client;
    $result=$client->getBucketAcl(array(
        'Bucket' => $bucketname
    ));
    echo "\n";
    echo "<br><br>Owner DisplayName: " . $result['Owner']['DisplayName'] . "<br>";
    echo "Owner ID: " . $result['Owner']['ID'] . "<br>";
    foreach ($result['Grants'] as $Grantee) {
        if (! empty($Grantee['Grantee']['Type'])) {
            echo  "Grantee type: " . $Grantee['Grantee']['Type'] . "<br>";
        }
        if (! empty($Grantee['Grantee']['ID'])) {
            echo  "Grantee ID: " . $Grantee['Grantee']['ID'] ."<br>";
        }
        if (! empty($Grantee['Grantee']['URI'])) {
            echo  "Grantee URI: " . $Grantee['Grantee']['URI'] ."<br>";
        }
        if (! empty($Grantee['Grantee']['DisplayName'])) {
            echo  "Grantee DisplayName: " . $Grantee['Grantee']['DisplayName'] . "<br>";
        }
        if (! empty($Grantee['Permission'])) {
            echo  "Grantee Permission: " . $Grantee['Permission']."<br><br>";
        }
    }
    echo "\n";
}

function BasicputBucket()
{
    global $bucketname;
    global $client;
    global $userAEmail;
    global $ownerCanonicalId;
    
    //CreateBucket
    createBucket($bucketname);
    //SetBucketACL
    $client->putBucketAcl(array(
        'Bucket' => $bucketname ,
        'Grants' => [
            [
                'Grantee' => [
                    'ID' => "{$ownerCanonicalId}",
                    'Type' => 'CanonicalUser',
                ],
                'Permission' => 'FULL_CONTROL',
            ],
            [
                'Grantee' => [
                    'EmailAddress' => "{$userAEmail}",
                    'Type' => 'AmazonCustomerByEmail',
                ],
                'Permission' => 'FULL_CONTROL',
            ],
            [
                'Grantee' => [
                    'Type' => 'Group',
                    'URI' => 'http://acs.amazonaws.com/groups/global/AllUsers'
                ],
                'Permission' => 'FULL_CONTROL',
            ],
        ],
        'Owner' => [
            'ID' => "{$ownerCanonicalId}"
        ]
    ));
    
    //SetBucketACL by Canned ACL
    $client->putBucketAcl(array(
            'Bucket' => $bucketname ,
            'ACL'	 => 'public-read'
    ));
    
    showBucketACL($bucketname);
    //DeleteBucket
    $client->deleteBucket(array('Bucket' => $bucketname));
}

function putBucketV()
{
    global $bucketname;
    global $client;
    global $userAEmail;
    global $ownerCanonicalId;
        
    //CreateBucket
    $file="testv.txt";
    createBucket($bucketname);
        
    //Enable Bucket Version
    $client->putBucketVersioning(array(
        'Bucket' => $bucketname,
        'Status' => 'Enabled'
    ));
    
    //PutObject
    putObject($bucketname, $file, createSampleFile());
    
    //GetObject
    $result=$client->getObject(array(
        'Bucket' => $bucketname,
        'Key'	 => $file
    ));
        
    //SetObjectACL
    $client->putObjectAcl(array(
        'Bucket' => $bucketname ,
        'Key'	 => $file ,
        'Grants' => [
            [
                'Grantee' => [
                    'ID' => "{$ownerCanonicalId}",
                    'Type' => 'CanonicalUser',
                ],
                'Permission' => 'FULL_CONTROL',
            ],
            [
                'Grantee' => [
                    'EmailAddress' => "{$userAEmail}",
                    'Type' => 'AmazonCustomerByEmail',
                ],
                'Permission' => 'FULL_CONTROL',
            ],
            [
                'Grantee' => [
                    'Type' => 'Group',
                    'URI' => 'http://acs.amazonaws.com/groups/global/AllUsers'
                ],
                'Permission' => 'FULL_CONTROL',
            ],
        ],
        'Owner' => [
            'ID' => "{$ownerCanonicalId}"
        ],
    ));

    showObjectACL($bucketname, $file);
        
    //List all versions in your bucket
    $result=$client->listObjectVersions(array(
                'Bucket' => $bucketname
        ));
        
    $result = $client->listObjectVersions(array(
            'Bucket' => $bucketname,
        ));
    $i=0;
    foreach ($result['Versions'] as $v) {
        $a[$i] = $v['VersionId'];
        $i++;
    }
        
    //Delete Objects
    $objs = array();
    for ($i=0;$i<count($a);$i++) {
        array_push($objs, array( 'Key' => $file, 'VersionId' => $a[$i]));
    }

    if (count($objs) > 0) {
        $result = $client->deleteObjects(array(
            'Bucket' => $bucketname,
            'Objects' => $objs
        ));
    }
        
    //DeleteBucket
    $client->deleteBucket(array('Bucket' => $bucketname));
}

$bucketname=$argv[1];
$userAEmail=$argv[4];
$ownerCanonicalId=$argv[5];

try {
    echo "ACLSerialTesting...";
    /*
     * test 1. create Bucket
     * 		2. put canned ACL & ACP
     * 		2. get Bucket List
     * 		3. get Bucket ACL & ACP
     *
     */
    BasicputBucket();
    /*
     * test 1. create Bucket
     * 		2. put Bucket versioning
     * 		3. put object
     * 		4. get object (version ID)
     * 		5. put canned ACL & ACP
     */
    putBucketV();
} catch (S3Exception $e) {
    echo "Caught an AmazonServiceException.", "\n";
    echo "Error Message:    " . $e->getMessage(). "\n";
    echo "HTTP Status Code: " . $e->getStatusCode(). "\n";
    echo "AWS Error Code:   " . $e->getExceptionCode(). "\n";
    echo "Error Type:       " . $e->getExceptionType(). "\n";
    echo "Request ID:       " . $e->getRequestId(). "\n";
}
